### PR TITLE
Fixes #12970 - Package patternfly-sass

### PR DESCRIPTION
**Pull request by dLobatog**
* [x] Add patternfly-sass
* [x] Update bootstrap-sass to 3.3.6 - https://github.com/theforeman/foreman-packaging/pull/951
* [x] Add autoprefixer-rails 5.2.1.3 - https://github.com/theforeman/foreman-packaging/pull/950
* [x] Update sprockets-rails to 2.3.3 - https://github.com/theforeman/foreman-packaging/pull/949
* [x] Add sprockets 3 (at least). - https://github.com/theforeman/foreman-packaging/pull/954
* [x] Add ejs 1.1 for sprockets 3 - https://github.com/theforeman/foreman-packaging/pull/965
* [x] Add sass-rails 5.0, ror41 max version is 4.0.3 - https://github.com/theforeman/foreman-packaging/pull/961
* [x] Add font-awesome-sass - #943
* [x] Add sass 3.4.19 - https://github.com/theforeman/foreman-packaging/pull/966
* [x] Update foreman.spec to depend on the aforementioned packages
* [x] Build a temporary repository with this 
[patternfly.zip](https://github.com/theforeman/foreman-packaging/files/91820/patternfly.zip)

cc - theforeman/foreman#294

### CODE REVIEW

File: rubygem-patternfly-sass/rubygem-patternfly-sass.spec [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+%{!?scl:%global pkg_name %{name}}
+
+%global gem_name patternfly-sass
+
+Name: %{?scl_prefix}rubygem-%{gem_name}
+Version: 2.8.0
+Release: 1%{?dist}
+Summary: Red Hat's Patternfly, converted to Sass and ready to drop into Rails
+Group: Development/Languages
+License: Apache-2.0
```
Comment (domcleal) [inline rubygem-patternfly-sass/rubygem-patternfly-sass.spec]: `ASL 2.0` in a spec file

File: rubygem-patternfly-sass/rubygem-patternfly-sass.spec [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+%build
+
+%install
+mkdir -p %{buildroot}%{gem_dir}
+cp -a .%{gem_dir}/* \
+        %{buildroot}%{gem_dir}/
+
+
+
```
Comment (domcleal) [inline rubygem-patternfly-sass/rubygem-patternfly-sass.spec]: remove some blank lines here please

File: rubygem-patternfly-sass/rubygem-patternfly-sass.spec [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+# Run the test suite
+%check
```
Comment (domcleal) [inline rubygem-patternfly-sass/rubygem-patternfly-sass.spec]: ideally needs filling in to run whatever specs are in the project, else it should be removed

File: rubygem-patternfly-sass/rubygem-patternfly-sass.spec [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+pushd .%{gem_instdir}
+
+popd
+
+%files
+%dir %{gem_instdir}
+%exclude %{gem_instdir}/.gitignore
+%{gem_instdir}/.rspec
+%exclude %{gem_instdir}/.rubocop.yml
+%exclude %{gem_instdir}/.travis.yml
```
Comment (domcleal) [inline rubygem-patternfly-sass/rubygem-patternfly-sass.spec]: You may want to just exclude `.../.*`

File: rubygem-patternfly-sass/rubygem-patternfly-sass.spec [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+%license %{gem_instdir}/LICENSE.txt
```
Comment (domcleal) [inline rubygem-patternfly-sass/rubygem-patternfly-sass.spec]: %license should be %doc for EL6 rpm compatibility

File: rubygem-patternfly-sass/rubygem-patternfly-sass.spec [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+%files doc
+%doc %{gem_docdir}
+%{gem_instdir}/Gemfile
+%doc %{gem_instdir}/README.md
+%{gem_instdir}/Rakefile
+%{gem_instdir}/patternfly-sass.gemspec
+%{gem_instdir}/spec
+
+%changelog
+* Thu Dec 31 2015 Daniel Lobato - 2.8.0-1
```
Comment (domcleal) [inline rubygem-patternfly-sass/rubygem-patternfly-sass.spec]: missing e-mail address

**Comment by domcleal** [comment]
1) as per #943, missing comps/ and rel-eng/ updates.  Please rebase first to avoid a conflict.
2) missing runtime requires
3) missing relevant versions of sass and bootstrap-sass, they'd also need updating

**Comment by domcleal** [comment]
It looks like you're updating a lot of packages to fully implement the changes in #4629, so you might find this easier in one PR and/or with a temporary repo (e.g. using mockchain or copr) to show the changes work together - including the changes to foreman/foreman.spec.

**Comment by dLobatog** [comment]
It looks like I have to:

* Add patternfly-sass (done?)
* Update bootstrap-sass to 3.3.6
* Add autoprefixer-rails 5.2.1.3
* Update sprockets-rails to 2.3.3 (at least). 2.2.4 is the maximum version in [the ror41 SCL](https://www.softwarecollections.org/repos/rhscl/rh-ror41/epel-6-x86_64/), would it be worth it to submit changes there directly?
* Add sprockets 3 (at least). The maximum version on ror41 is 2.12.3.
* Add sass-rails 5.0, ror41 max version is 4.0.3
* Add font-awesome-sass - https://github.com/theforeman/foreman-packaging/pull/943
* Update foreman.spec to depend on the aforementioned packages

I think that would be it, I'll add it to the original comment as a check list. @domcleal, [you said](https://github.com/theforeman/foreman-packaging/pull/944#issuecomment-168624379) you would prefer all of these changes in this PR, plus a copr with the packages then? Most of the items in the list above don't necessarily have to rely on one another so they could be merged independently.

**Comment by domcleal** [comment]
> Update sprockets-rails to 2.3.3 (at least). 2.2.4 is the maximum version in the ror41 SCL, would it be worth it to submit changes there directly?

I doubt very much that you could get this changed in the rh-ror41 SCL, since it's a rebuild of a Red Hat product, so the "upstream" has a policy of no rebases.

I don't mind if this is in one PR or many, but I think the series of packages will need to be chainbuilt to prove that it all works together (mockchain with the configs in mock/ and SRPMs from `tito build --srpm --test` should be easy).

**Comment by domcleal** [comment]
Don't worry about Redmine tickets for each change too, it's not necessary here - just more admin work.

File: comps/comps-foreman-fedora21.xml [hunk 7]
```diff
@@ -140,6 +140,7 @@
       <packagereq type="default">rubygem-multi_json-doc</packagereq>
       <packagereq type="default">rubygem-multi-select-rails-doc</packagereq>
       <packagereq type="default">rubygem-net-ldap-doc</packagereq>
+      <packagereq type="default">rubygem-patternfly-sass-doc</packagereq>
```
Comment (domcleal) [inline comps/comps-foreman-fedora21.xml]: missing the main package?

File: foreman/foreman.spec [hunk 8]
```diff
@@ -190,6 +190,8 @@ BuildRequires: %{?scl_prefix_ror}rubygem(execjs) >= 1.4.0
 BuildRequires: %{?scl_prefix_ror}rubygem(execjs) < 2.5.0
 BuildRequires: %{?scl_prefix_ror}rubygem(jquery-rails) >= 3.1
 BuildRequires: %{?scl_prefix_ror}rubygem(jquery-rails) < 4.0
+BuildRequires: %{?scl_prefix_ror}rubygem(patternfly-sass) >= 2.7
```
Comment (domcleal) [inline foreman/foreman.spec]: scl_prefix, not in the ror collection

Comment (domcleal) [inline foreman/foreman.spec]: Nitpick, if you can move it below jquery-ui-rails then it keeps the layout the same as the assets.rb file - making it much quicker to review changes.

File: foreman/foreman.spec [hunk 9]
```diff
@@ -368,6 +370,8 @@ Requires: %{?scl_prefix_ror}rubygem(uglifier) >= 1.0.3
 Requires: %{?scl_prefix_ror}rubygem(execjs) >= 1.4.0
 Requires: %{?scl_prefix_ror}rubygem(jquery-rails) >= 3.1
 Requires: %{?scl_prefix_ror}rubygem(jquery-rails) < 4.0
+Requires: %{?scl_prefix_ror}rubygem(patternfly-sass) >= 2.7
```
Comment (domcleal) [inline foreman/foreman.spec]: ditto both comments above

File: rubygem-patternfly-sass/rubygem-patternfly-sass.spec [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+Group: Development/Languages
+License: ASL 2.0
+URL: https://github.com/Patternfly/patternfly-sass
+Source0: http://rubygems.org/gems/%{gem_name}-%{version}.gem
+Requires: %{?scl_prefix_ror}rubygem(bootstrap-sass) >= 3.3.5
+Requires: %{?scl_prefix_ror}rubygem(bootstrap-sass) < 3.4.0
+Requires: %{?scl_prefix_ror}rubygem(font-awesome-sass) >= 4.3.0
+Requires: %{?scl_prefix_ror}rubygem(font-awesome-sass) < 4.4.0
+Requires: %{?scl_prefix_ror}rubygem(sass) > 3.4.15
+Requires: %{?scl_prefix_ror}rubygem(sass) < 3.5.0
```
Comment (domcleal) [inline rubygem-patternfly-sass/rubygem-patternfly-sass.spec]: no _ror on any of these dependencies, as you're adding them into our SCL collection

File: rubygem-patternfly-sass/rubygem-patternfly-sass.spec [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+Release: 1%{?dist}
+Summary: Red Hat's Patternfly, converted to Sass and ready to drop into Rails
+Group: Development/Languages
+License: ASL 2.0
+URL: https://github.com/Patternfly/patternfly-sass
+Source0: http://rubygems.org/gems/%{gem_name}-%{version}.gem
+Requires: %{?scl_prefix_ror}rubygem(bootstrap-sass) >= 3.3.5
+Requires: %{?scl_prefix_ror}rubygem(bootstrap-sass) < 3.4.0
+Requires: %{?scl_prefix_ror}rubygem(font-awesome-sass) >= 4.3.0
+Requires: %{?scl_prefix_ror}rubygem(font-awesome-sass) < 4.4.0
```
Comment (domcleal) [inline rubygem-patternfly-sass/rubygem-patternfly-sass.spec]: no _ror on these bootstrap/font-awesome dependencies, as you're adding them into our SCL collection

File: rubygem-patternfly-sass/rubygem-patternfly-sass.spec [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+Requires: %{?scl_prefix_ror}rubygem(sass) > 3.4.15
```
Comment (domcleal) [inline rubygem-patternfly-sass/rubygem-patternfly-sass.spec]: `>=`

File: comps/comps-foreman-fedora21.xml (hunk 7 above)
Comment (dLobatog) [inline comps/comps-foreman-fedora21.xml]: It's on line 75

**Comment by domcleal** [comment]
Please let me know when you have a set of packages tested together.  It should be possible to use mockchain with the configs provided in the mock/ directory to build all of these packages in sequence and then the SRPM of foreman.

At the moment they're also going to have quite a few merge conflicts in foreman.spec, which'll make it a bit trickier.

**Comment by dLobatog** [comment]
I'm building this temporary repo now with:
```bash

for dir in rubygem-*; do 
(cd "$dir" && 
tito build --srpm --test\
 --builder tito.builder.MockBuilder\
 --arg mock_config_dir=mock/ --arg mock=el7-scl); 
done

mockchain -r el7-scl -l /tmp/patternfly/ --recurse /tmp/tito/*.src.rpm
```

**Comment by dLobatog** [comment]
Duh, it looks like rubygem(ejs) is not packaged. I'll take it from https://github.com/domcleal/sclo-ror42/tree/master/rubygem-ejs

**Comment by dLobatog** [comment]
For the record I'm keeping a branch with all of the [changes here](https://github.com/dlobatog/foreman-packaging/tree/rpm/patternfly-all] 

If that builds properly (currently it doesn't), I'll port the relevant changes to every PR on the checklist.

**Comment by dLobatog** [comment]
Success! I'll start updating the PRs with the few changes I made, basically https://github.com/dLobatog/foreman-packaging/commit/ad77506b327ac2a1b6f79a0c785ffce674020218

**Comment by dLobatog** [comment]
Find the output of mockchain at [patternfly.zip](https://github.com/theforeman/foreman-packaging/files/90587/patternfly.zip)

**Comment by domcleal** [comment]
You should then be able to build foreman itself on top of that.  Get an SRPM with the latest develop code with `cd foreman && tito release --scratch --arg jenkins_job=test_develop koji-foreman-nightly --dry-run`, then run it on your existing chain repo.

**Comment by dLobatog** [comment]
It's not building now because rubygem(sass) from ror41 is also too old.
```
Error: Package: tfm-rubygem-patternfly-sass-2.8.0-1.git.0.7154183.el7.centos.noarch (local_build_repo)
           Requires: rh-ror41-rubygem(sass) >= 3.4.15
           Installing: rh-ror41-rubygem-sass-3.3.8-1.el7.centos.noarch (scl-ror)
               rh-ror41-rubygem(sass) = 3.3.8
```
I'll bundle it too.

**Comment by dLobatog** [comment]
@domcleal I built foreman with mockchain + the other packages. Find the zip here: [patternfly.zip](https://github.com/theforeman/foreman-packaging/files/91820/patternfly.zip)

I worked out of this branch https://github.com/dLobatog/foreman-packaging/commits/rpm/patternfly-all then split it all in all the various PRs you can see here.

**Comment by domcleal** [comment]
Thanks, all looks good except for one comment on sprockets I think it was.  I'm working through the PRs in this order to match the build order:

- [x] rubygem-autoprefixer-rails
- [ ] rubygem-bootstrap-sass
- [ ] rubygem-ejs
- [ ] rubygem-font-awesome-sass
- [ ] rubygem-patternfly-sass
- [ ] rubygem-sass
- [ ] rubygem-sass-rails
- [ ] rubygem-sprockets
- [ ] rubygem-sprockets-rails

Once the previous one's merged the next is almost certainly going to need a rebase.  Don't worry about rebasing all of them as it'll just generate PR test load.

**Comment by domcleal** [comment]
Could you rebase this please @dLobatog?

**Comment by dLobatog** [comment]
@domcleal Rebased, it only conflicted on foreman.spec w/ autoprefixer-rails

**Comment by domcleal** [comment]
Thanks @dLobatog, merged and building.

**Comment by domcleal** [comment]
The core Foreman package has been rebuilt and passed tests.  Thanks for the updates @dLobatog!

**Pull request merged**